### ISSUE TITLE
[InferAddressSpaces] asserts when adding `addrspacecast` between inttoptr/ptrtoint pair

### ISSUE THREAD

**Issue by chichunchen** [llvm/llvm-project#53665]
Code snippet that causing the assert:
```
  %r2 = load i64, i64 addrspace(5)* %"$$arg_ptr_acc_grid_t30_t88", align 8
  %r3 = inttoptr i64 %r2 to i8 addrspace(1)*
  %r4 = addrspacecast i8 addrspace(1)* %r3 to i8*
  %r5 = ptrtoint i8* %r4 to i64
  %r6 = inttoptr i64 %r5 to double**
  %r7 = load double*, double** %r6, align 8
```
If we remove `%r4` and replace the use of `%r4` with %r3, then the assertion error disappear

Test case running on Godbolt: https://godbolt.org/z/E8vjKhj8M

Error message from llc:
```
llc: /root/llvm-project/llvm/lib/Transforms/Scalar/InferAddressSpaces.cpp:658: llvm::Value* {anonymous}::InferAddressSpacesImpl::cloneInstructionWithNewAddressSpace(llvm::Instruction*, unsigned int, const ValueToValueMapTy&, const PredicatedAddrSpaceMapTy&, llvm::SmallVectorImpl<const llvm::Use*>*) const: Assertion `Src->getType()->getPointerAddressSpace() == NewAddrSpace' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: /opt/compiler-explorer/clang-assertions-trunk/bin/llc -o /app/output.s -x86-asm-syntax=intel <source>
1.	Running pass 'Function Pass Manager' on module '<source>'.
2.	Running pass 'Infer address spaces' on function '@func'
 #0 0x00005628b552e29f PrintStackTraceSignalHandler(void*) Signals.cpp:0:0
 #1 0x00005628b552ba5d SignalHandler(int) Signals.cpp:0:0
 #2 0x00007fae42ceb3c0 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x153c0)
 #3 0x00007fae427ba18b raise (/lib/x86_64-linux-gnu/libc.so.6+0x4618b)
 #4 0x00007fae42799859 abort (/lib/x86_64-linux-gnu/libc.so.6+0x25859)
 #5 0x00007fae42799729 (/lib/x86_64-linux-gnu/libc.so.6+0x25729)
 #6 0x00007fae427aaf36 (/lib/x86_64-linux-gnu/libc.so.6+0x36f36)
 #7 0x00005628b4f0bf4e (anonymous namespace)::InferAddressSpacesImpl::cloneInstructionWithNewAddressSpace(llvm::Instruction*, unsigned int, llvm::ValueMap<llvm::Value const*, llvm::WeakTrackingVH, llvm::ValueMapConfig<llvm::Value const*, llvm::sys::SmartMutex<false> > > const&, llvm::DenseMap<std::pair<llvm::Value const*, llvm::Value const*>, unsigned int, llvm::DenseMapInfo<std::pair<llvm::Value const*, llvm::Value const*>, void>, llvm::detail::DenseMapPair<std::pair<llvm::Value const*, llvm::Value const*>, unsigned int> > const&, llvm::SmallVectorImpl<llvm::Use const*>*) const (.isra.413) InferAddressSpaces.cpp:0:0
 #8 0x00005628b4f0ed6c (anonymous namespace)::InferAddressSpacesImpl::run(llvm::Function&) (.constprop.518) InferAddressSpaces.cpp:0:0
 #9 0x00005628b4f110f6 (anonymous namespace)::InferAddressSpaces::runOnFunction(llvm::Function&) InferAddressSpaces.cpp:0:0
#10 0x00005628b4d32816 llvm::FPPassManager::runOnFunction(llvm::Function&) (/opt/compiler-explorer/clang-assertions-trunk/bin/llc+0x2932816)
#11 0x00005628b4d32b69 llvm::FPPassManager::runOnModule(llvm::Module&) (/opt/compiler-explorer/clang-assertions-trunk/bin/llc+0x2932b69)
#12 0x00005628b4d334a9 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/opt/compiler-explorer/clang-assertions-trunk/bin/llc+0x29334a9)
#13 0x00005628b2f56578 compileModule(char**, llvm::LLVMContext&) llc.cpp:0:0
#14 0x00005628b2e94adf main (/opt/compiler-explorer/clang-assertions-trunk/bin/llc+0xa94adf)
#15 0x00007fae4279b0b3 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b3)
#16 0x00005628b2f4d03a _start (/opt/compiler-explorer/clang-assertions-trunk/bin/llc+0xb4d03a)
Compiler returned: 139
```

**Comment by arsenm** [comment]
https://reviews.llvm.org/D119306

**Comment by arsenm** [comment]
Should be fixed by 52fbb786a638ecc7349641b45b62a5abafffdf75
